brainclouds2s.js

```diff
@@ -241,31 +188,44 @@ function request(context, json, callback)
             return
         }
 
-        context.requestQueue.splice(0, 1); // Remove this request
+        if (data && data.messageResponses && data.messageResponses.length > 0) {
+            callCallback(callback, context, data.messageResponses[0])
+        }
+        else {
+            
+            // Error that has no packets
+            callCallback(callback, context, data)
+        }
+
+        // This request is complete and safe to remove from queue after checking for callback
+        context.requestQueue.splice(0, 1); // Remove this request from the queue
         context.retryCount = 0;
 
-        if (callback)
-        {
-            if (data && data.messageResponses && data.messageResponses.length > 0)
-            {
-                callback(context, data.messageResponses[0])
-            }
-            else
-            {
-                // Error that has no packets
-                callback(context, data)
-            }
-        }
-        
         // Do next request in queue
-        if (context.requestQueue.length > 0)
-        {
+        if (context.requestQueue.length > 0) {
             let nextRequest = context.requestQueue[0];
             request(context, nextRequest.json, nextRequest.callback)
         }
     })
 }
 
+/**
+ * Invoke a given callback.
+ * @param callback Function to be executed
+ * @param context S2S context object
+ * @param data Content object to be sent
+ */
+function callCallback(callback, context, data) {
+    if (callback != null) {
+        try {
+            callback(context, data);
+        }
+        catch (error) {
+            console.log(`[S2S Error ${context.appId}] ${error}`)
```

Comment:
Having thought about this some more, I think we should be a little more specific with the error message here... something like "[S2S Error in callback ..."